climada/hazard/base.py

```diff
@@ -191,6 +191,32 @@ def __init__(self,
         if self.pool:
             LOGGER.info('Using %s CPUs.', self.pool.ncpus)
 
+    @property
+    def intensity(self) -> sparse.csr_matrix:
+        """Hazard intensity matrix"""
+        return self._intensity
+
+    @intensity.setter
+    def intensity(self, value: sparse.csr_matrix):
+        """Set intensity matrix to new value"""
+        self._intensity = value
+        self._intensity.check_format()
+        self._intensity.eliminate_zeros()
+        self._intensity.sum_duplicates()
```

Comment:
I am not fully convinced that we should do that silently. Can we not instead raise an error if the input `crs_matrix` has duplicates? Maybe the property `crs_matrix.has_canonical_format` would be the thing to check?

Reply:
Two reasons against this:
1. Several tests fail if we introduce an explicit check for `has_canonical_format`. I did not dig deeper to check why exactly. But it is not uncommon throughout the Climada code base.
2. A non-canonical matrix is completely fine from the implementation perspective. It just has two disadvantages I see:
   * `.data` does not reflect the actual matrix values. I think this is irrelevant for 90% of all users, they will never touch that attribute.
   * Accessing matrix values might be slower if indices are not ordered and there are duplicates. Not sure how much, though.

The overhead for eliminating zeros and summing duplicates should be very low, especially since it will speed up the impact calculation in turn. Because no "visible" data of the matrix is changed, I think silently calling these methods is fine

Reply:
hmmm ok. Then maybe we should add it as information in the docstring of the class?

Reply:
Yes, can do.

Come to think of it, we should probably also call these checks when the Hazard is plugged into `ImpactCalc`, because direct modifications of the csr_matrix might again produce duplicates or explicit zeros.